[config] Add abbitily to deprecate a node

sanpii opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | maybe
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/7794

For BC breaks, I don’t know if modifying the Xml and Yaml dumper output is considering as a BC break (it’s just a comment).

sanpii committed:
[config] Add abbitily to deprecate a node

Nek- commented:
You may add the possibility to specify a message ? Anyway that's awesome, thanks !

HeahDude commented:
Hello @sanpii, thanks for proposing this. For the record, this was proposed before https://github.com/symfony/symfony/pull/21051.

sanpii commented:
> You may add the possibility to specify a message ?

It’s in a corner of my mind, but I don’t know how do this smartly.

sstok commented:
FYI, there is typo in the title `depreciate` -> `deprecate` 😉
depreciate is to devalue something.

sanpii commented:
@sstok fixed, thank you.

ro0NL commented:
To clarify, my goal with #21051 was to _not_ include/define deprecated nodes in the tree; and solely focus on bridging the old situation to a new situation.

Im :-1: on this. You should checkout why this was rejected :)

sanpii commented:
> You should checkout why mine was rejected :)

I see two points:

> is quite easily implemented directly in the code

Of course, all new features is probably easy to implement… But, I think, `trigger_error` is a low level function and symfony is here to abstract its.

> I don't see the need for end user projects to deprecate options

We are, at least, two persons who find this suffisamment helpful to take the time to make a PR.

ro0NL commented:
I still think it's a good feature.. in case you thought different.

Maybe this approach is good enough, as it's kinda the same we do for services. Personally i like the normalization to happen, otherwise people still need to take care of old/new nodes at implementation level, e.g. in an extension.

nicolas-grekas commented on src/Symfony/Component/Config/Definition/ArrayNode.php:
```diff
@@ -234,6 +234,11 @@ protected function finalizeValue($value)
 
         foreach ($this->children as $name => $child) {
+            if ($child->isDeprecated()) {
+                $msg = sprintf('The child node "%s" at path "%s" is deprecated.', $name, $this->getPath());
```
Typically, deprecation message should also mention the version since it's been deprecated, the outcome in the next major version, and the alternative if any. Would be great to allow + encourage doing so here also.

sanpii replied:
Yes, I’m agree  with you. But I did’t find a good way to do this.

Maybe by passing an optionnal argument to `isDeprecated()` method?

Taluu replied:
I think that customizing the message would solve the thing (with mandatory place holders like `%node_name%` and `%node_path%`. A bit like what is done on deprecating services, which has a mandatory `%service_id%` placeholder (or something like that).

nicolas-grekas commented:
I think we can go with this feature, the implementation is simple enough.
I propose this API:
```
public function setDeprecated($message)
public function isDeprecated()
public function getDeprecationMessage($node, $path)
```

with `$message` allowing `%node%` and `%path` placeholders.

Status: needs work

sanpii commented:
PR updated according to @nicolas-grekas’ propositions.

robfrawley commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -182,6 +193,29 @@ public function isRequired()
+     */
+    public function getDeprecationMessage($node, $path)
+    {
+        return strtr($this->deprecatedMessage, array('%node%' => $node, '%path%' => $path));
```
Isn't `sprintf` (or, in your case, if you'd like to pass an array, `vsprintf`) significantly faster than the other string replacement functions? See, for example: https://www.simplemachines.org/community/index.php?topic=175031.0

ro0NL commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -142,6 +143,16 @@ public function setRequired($boolean)
     /**
+     * Set this node as deprecated.
+     *
+     * @param string $message Deprecated message
```
`string|null`

ro0NL commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -182,6 +193,29 @@ public function isRequired()
+     */
+    public function isDeprecated()
+    {
+        return $this->deprecatedMessage != null;
```
`!==`

ro0NL commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -182,6 +193,29 @@ public function isRequired()
+     *
+     * @return bool
+     */
+    public function getDeprecationMessage($node, $path)
```
`protected`?

sanpii replied:
Not now, it’s used by Xml and Yaml dumper.

ro0NL commented on src/Symfony/Component/Config/Tests/Fixtures/Configuration/ExampleConfiguration.php:
```diff
@@ -35,6 +35,8 @@ public function getConfigTreeBuilder()
                 ->scalarNode('scalar_array_defaults')->defaultValue(array('elem1', 'elem2'))->end()
                 ->scalarNode('scalar_required')->isRequired()->end()
+                ->scalarNode('scalar_deprecated')->isDeprecated()->end()
+                ->scalarNode('scalar_deprecated_with_message')->isDeprecated('Deprecation custom message')->end()
```
you should test the placeholder replacement as well imo.

sanpii commented:
@ro0NL Fixed, thank you for your review.

nicolas-grekas commented on src/Symfony/Component/Config/Definition/ArrayNode.php:
```diff
@@ -234,6 +234,11 @@ protected function finalizeValue($value)
 
         foreach ($this->children as $name => $child) {
+            if ($child->isDeprecated()) {
+                $msg = $child->getDeprecationMessage($name, $this->getPath());
```
let get rid of the $msg var and do things inline

nicolas-grekas commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -182,6 +193,29 @@ public function isRequired()
+     */
+    public function isDeprecated()
+    {
+        return $this->deprecatedMessage !== null;
```
yoda style: `null !== ...`

nicolas-grekas commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -29,6 +29,7 @@
     protected $finalValidationClosures = array();
     protected $allowOverwrite = true;
     protected $required = false;
+    protected $deprecatedMessage = null;
```
`$deprecationMessage` (same below)

nicolas-grekas commented on src/Symfony/Component/Config/CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@ CHANGELOG
  * made `ClassExistenceResource` work with interfaces and traits
  * added `ConfigCachePass` (originally in FrameworkBundle)
  * added `castToArray()` helper to turn any config value into an array
+ * added `isDeprecated()` method to indicate a deprecated node
```
let advertise `setDeprecated` instead?

nicolas-grekas commented on src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php:
```diff
@@ -169,6 +170,20 @@ public function isRequired()
+     *
+     * @return $this
+     */
+    public function isDeprecated($message = 'The child node "%node%" at path "%path%" is deprecated.')
```
should be named `setDeprecated`.
Not sure about the default value, there is always more to add than the default (not a string opinion, just noting.)

nicolas-grekas commented on src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php:
```diff
@@ -169,6 +170,20 @@ public function isRequired()
+     */
+    public function isDeprecated($message = 'The child node "%node%" at path "%path%" is deprecated.')
+    {
+        $this->deprecatedMessage = $message;
```
[in Definition](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Definition.php#L680), we enforce that the placeholder is used, we should do the same here IMHO, for both placeholders

sanpii commented:
@nicolas-grekas thank you.

Nyholm committed:
[Translation] Adding the ability do load <notes> in xliff2.0

fabpot committed:
feature #23947 [Translation] Adding the ability do load <notes> in xliff2.0 (Nyholm)

This PR was squashed before being merged into the 3.4 branch (closes #23947).

Discussion
----------

[Translation] Adding the ability do load <notes> in xliff2.0

| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | ~~~~yes~~~ no (sorry)
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

Since #23890 we can dump `<notes>`. We should also have the ability to load them back.

Commits
-------

b0cdb53d67 [Translation] Adding the ability do load <notes> in xliff2.0

fabpot commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -142,6 +143,16 @@ public function setRequired($boolean)
     }
 
     /**
+     * Set this node as deprecated.
```
`Sets`

fabpot commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -182,6 +193,29 @@ public function isRequired()
+     * @param string $node the configuration node name
+     * @param string $path the path of the node
+     *
+     * @return bool
```
`@return string`

fabpot commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -142,6 +143,16 @@ public function setRequired($boolean)
 
     /**
+     * Set this node as deprecated.
+     *
```
You should hint here that `%node%` and `%path%` have a special meaning in a deprecated message.

fabpot commented on src/Symfony/Component/Config/Definition/Builder/NodeDefinition.php:
```diff
@@ -169,6 +170,20 @@ public function isRequired()
 
     /**
+     * Sets the node as deprecated.
+     *
```
Same here for the special %node% and %path% vars.

fabpot commented on src/Symfony/Component/Config/Definition/Dumper/XmlReferenceDumper.php:
```diff
@@ -153,6 +153,13 @@ private function writeNode(NodeInterface $node, $depth = 0, $root = false, $name
+                        $comments[] = sprintf(
+                            'Deprecated (%s)',
+                            $child->getDeprecationMessage($child->getName(), $child->getPath())
+                        );
```
should be on one line.

fabpot commented on src/Symfony/Component/Config/Definition/Dumper/YamlReferenceDumper.php:
```diff
@@ -123,6 +123,14 @@ private function writeNode(NodeInterface $node, $depth = 0, $prototypedArray = f
+            $comments[] = sprintf(
+                'Deprecated (%s)',
+                $node->getDeprecationMessage($node->getName(), $node->getPath())
+            );
```
one line please

fabpot commented:
Target branch should also be changed to 3.4 instead of master.

sanpii commented:
@fabpot fixed and rebased on the 3.4 branch. Thank you.

fabpot commented:
Thank you @sanpii.

stof commented:
@fabpot why was it merged into the 4.0-dev branch ?

nicolas-grekas commented:
@sanpii are there existing deprecated options that could leverage this PR (on 3.4)? Would you have time to create a PR for it?

maidmaid commented:
https://github.com/symfony/symfony/blob/4173f1389ec935caa34396f17a1cd32b602c7af0/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L68-L76
Maybe here for instance?

sanpii commented:
@nicolas-grekas good idea, but the PR was merged on master, not on 3.4

Here the deprecated configuration I found:

https://github.com/symfony/symfony/blob/3.4/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L556
https://github.com/symfony/symfony/blob/3.4/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L1199
https://github.com/symfony/symfony/blob/3.4/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L1455
https://github.com/symfony/symfony/blob/3.4/src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php#L67

nicolas-grekas commented:
Now merged in 3.4
